apps/meteor/server/services/authorization/canAccessRoom.ts

```diff
@@ -89,7 +109,9 @@ export const canAccessRoom: RoomAccessValidator = async (room, user, extraData):
 
 	for await (const roomAccessValidator of roomAccessValidators) {
 		if (await roomAccessValidator(room, user, extraData)) {
-			return true;
+			if (await hasRoomViewPermission(room.t, user._id)) {
```

Comment:
There's no need to run this inside the validator loop; Instead, run it first and if there's no permission, return false;

Also: please double-check if this won't break livechat visitor access. I think it will.